### ISSUE TITLE
Coverage tool for scalatest

### ISSUE THREAD

**Issue by johanneshiry** [scalatest/scalatest#1823]
Hi everyone,

we're using scalatest in one of our major projects and we really enjoy it's capabilities. Thanks for the good work! Anyway we're missing support of any coverage report tool we know for scalatest. I spend some time googling around to find some but without success. 

Do you know if there is any tool outside that is able to generate coverage reports based on the scalatest results? Preferably with gradle support. 

Thanks in advance for your help!

Best,

Johannes

**Comment by antoine-mulet** [comment]
Hi @johanneshiry ,

Your question caught my eye because I recently spent some time playing with coverage tools in a project where we use ScalaTest. My understanding is that a testing tool like ScalaTest and code coverage tools are in fact two very different things. While the testing framework lets you write and run your tests, the code coverage tool simply records the parts of you production code that have been hit by your tests runs ultimately to produce coverage reports. So basically what I am saying is that they are independent and you usually "glue" them together using a build tool like gradle to be able to run your tests, records the coverage and produce the coverage reports with a single command.

I assume that you are looking for a code coverage tool for Scala code given you are using ScalaTest (although the framework supports Scala, Scala.js (JavaScript), and Java code according to the documentation). If that's the case then you should have a look at http://scoverage.org/. I have used it with sbt but it seems to have gradle support as well via this [plugin](https://github.com/scoverage/gradle-scoverage). Otherwise you can also have a look at [Jacoco](https://www.eclemma.org/jacoco/). It is not Scala specific but ultimately because it's bytecode-based it also works for Scala. I personally tried both and I found scoverage coverage reports a bit more accurate simply because it has been designed specifically for Scala. As I mentioned above I tried both using sbt and ScalaTest but I am pretty sure the same can be achieved using gradle.

I hope it helps.

Antoine

**Comment by johanneshiry** [comment]
Hi @antoine-mulet,
thanks for your reply and the extensive explanation on test suits and coverage tools. 
I've a little bit of experience in this field, as I'm responsible for several CI/CD pipelines in which make use of different test suits (JUnit, ScalaTest, Spock) combined with different coverage tools for analysis (mainly Jacoco which is passed to sonarqube instances). 

Considering your proposal of using scoverage (which I really, really like), I have to say that I tried "glueing" together gradle-scoverage with ScalaTest, but gave up at same point. 
The first issue has been a lacking support of scala 2.13 (see my issue [109](https://github.com/scoverage/gradle-scoverage/issues/109)) and then some issues with multi-project builds (issue [98](https://github.com/scoverage/gradle-scoverage/issues/98)), while the multi-project issue is in fact not a gradle-scoverage problem but it _seems_ to be a general gradle problem. 

After giving up on gradle-scoverage I tried using Jacoco with ScalaTest (because we have it running on several other java projects) but it also _seems_ to have issues with ScalaTest, because the resulting test data has not been analyzed by Jacoco. Maybe I was missing something there, but we couldn't managed to get it working ScalaTest.

**tl;dr:**
- We gave up on gradle-scoverage due to the 2 issues mentioned above and the comment of one of the gradle-scoverage devs ["it uses ScalaTest instead of JUnit, which I'm not sure is compatible with this plugin. (this plugin == gradle scoverage"](https://github.com/scoverage/gradle-scoverage/issues/98#issuecomment-568278076)
- We couldn't manage to setup a working combination of ScalaTest and Jacoco. 
- If you have a working setup with ScalaTest and Jacoco OR scoverage I would be glad if you could share this.

Thanks again for your help. Looking forward on your comments on this.

Johannes

**Comment by antoine-mulet** [comment]
Hi @johanneshiry,

It seems that you actually know way more than I do on this topic :) I thought you were seeking guidance about how to get started to set up code coverage with a fairly new/simple project but it's clearly not the case.

I have only little experience with gradle so I am afraid I cannot help you more on this front. However regarding scoverage potentially not being "compatible"with ScalaTest, I don't really get it conceptually. My basic understanding is that the two things are independent like I said in my first response. Once the code has been compiled scoverage instrumentation turned on, I don't understand how the testing framework used to write and run your tests could matter. Again I am very much not an expert on this topic so I may be wrong and would be happy to be corrected by someone more knowledgeable.

The only thing I could propose would be to share with you a simple Scala 2.13 **sbt** based example of scoverage working with ScalaTest. I understand that gradle seems to be a fairly strong requirement in your case so it may not be of much interest. Plus I have to admit that I didn't do anything magical to make it all work, I just followed the [ScalaTest with sbt](https://www.scalatest.org/user_guide/using_scalatest_with_sbt) and [sbt-scoverage](https://github.com/scoverage/sbt-scoverage) documentations + a few tricks to make it suit my use case where I need some test suites to run in parallel (not in a multi-project setup though).

Good luck with your difficult but worthy quest of getting your code covered!

Antoine

**Comment by johanneshiry** [comment]
Hi @antoine-mulet,

thanks for your reply. I think we have roughly the same understanding of testing framework and coverage frameworks. However, as this caused non-detected coverage in the past and based on the comment of one of the developers of gradle-scoverage I was _assuming_ that there might be circumstances where "communication" between testing and coverage might fail for some reasons. 

As I already stated in my previous post edit, I finally managed to get it working with the latest version of gradle-scoverage released a few days ago. Thanks for your offer for the sample project anyway! 

Wish a good start into the week. 

Johannes
